add Path options for SAMFileWriterFactory

jean-philippe-martin opened the pull request:
### Description

Add overloads to SAMFileWriterFactory that take `Path` as input instead of `File`. The old version that took `File` now just redirect to their `Path` equivalent, so there's the same amount of code to test.

The motivation for this change is to support Java NIO's pluggable file systems. This change (in combination with Google's NIO provider) allows one to directly write a SAM, BAM, or CRAM file to a Google Cloud Storage bucket. See eg. [GATK's matching issue](https://github.com/broadinstitute/gatk/issues/2422).

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary (not necessary)
- [x] *Is* backward compatible (does not break binary or source compatibility)

magicDGS commented on src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:
```diff
@@ -465,40 +564,41 @@ public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final boolean p
         if (createIndex) {
             if (!IOUtil.isRegularPath(outputFile)) {
-                log.warn("Cannot create index for CRAM because output file is not a regular file: " + outputFile.getAbsolutePath());
+                log.warn("Cannot create index for CRAM because output file is not a regular file: " + outputFile);
```
Would not be better to use `outputFile.toURI()` for easier debugging?

jean-philippe-martin replied:
Yes, done!

magicDGS commented on src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:
```diff
@@ -465,40 +564,41 @@ public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final boolean p
                 }
                 catch (final IOException ioe) {
-                    throw new RuntimeIOException("Error creating index file for: " + outputFile.getAbsolutePath()+ BAMIndex.BAMIndexSuffix);
+                    throw new RuntimeIOException("Error creating index file for: " + indexPath);
```
Maybe the URI will be useful also here...

jean-philippe-martin replied:
Yes, done.

magicDGS commented on src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:
```diff
@@ -465,40 +564,41 @@ public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final boolean p
         }
         catch (final IOException ioe) {
-            throw new RuntimeIOException("Error creating CRAM file: " + outputFile.getAbsolutePath());
+            throw new RuntimeIOException("Error creating CRAM file: " + outputFile.toString());
```
Maybe the URI will be useful also here...

jean-philippe-martin replied:
Agreed!

magicDGS commented on src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:
```diff
@@ -303,7 +342,7 @@ public SAMFileWriter makeSAMWriter(final SAMFileHeader header, final boolean pre
             else return ret;
         } catch (final IOException ioe) {
-            throw new RuntimeIOException("Error opening file: " + outputFile.getAbsolutePath());
+            throw new RuntimeIOException("Error opening file: " + outputPath);
```
URI?

jean-philippe-martin replied:
You got it!

magicDGS commented on src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:
```diff
@@ -241,25 +254,38 @@ public SAMFileWriter makeBAMWriter(final SAMFileHeader header, final boolean pre
-            final boolean createIndex = this.createIndex && IOUtil.isRegularPath(outputFile);
+            OutputStream os = IOUtil.maybeBufferOutputStream(Files.newOutputStream(outputPath), bufferSize);
+            if (createMd5File) os = new Md5CalculatingOutputStream(os, IOUtil.addExtension(outputPath,".md5"));
+            final BAMFileWriter ret = new BAMFileWriter(os, outputPath.toString(), compressionLevel, deflaterFactory);
```
URI string better?

jean-philippe-martin replied:
Sure. Done!

jean-philippe-martin commented on src/test/java/htsjdk/samtools/CRAMComplianceTest.java:
```diff
@@ -261,6 +261,39 @@ public void testBAMThroughCRAMRoundTrip() throws IOException, NoSuchAlgorithmExc
+        }
+
+        // write the BAM records to a temporary CRAM
+        final File tempCRAMFile = File.createTempFile("testBAMThroughCRAMRoundTrip", CramIO.CRAM_FILE_EXTENSION);
```
Excellent idea! I see we already depend on jimfs.

I'm also adding it to the test of `IOUtil.addExtension`, so we can test non-default-filesystem paths.

I also changed testCRAMThroughBAMRoundTrip so it saves to jimfs instead of the filesystem, that should speed things up a little bit.

This leaves only one test using files. I'm thinking of leaving it so we exercise that path through the code, though we can remove it if you feel strongly about it (these methods just call `.toPath()` and then redirect to the path code anyways).

magicDGS replied:
My suggestion was related to use at least in the changed test the jims filesystem, but anyway I think that some of the tests should still have `File`, which is still quite common.

jean-philippe-martin commented:
Thank you, excellent suggestions! I'm glad this code will be useful to you. Let me know what you think of this revised version!

jean-philippe-martin commented:
I see that storing the file name as a URL broke the tests; but that's a good thing because it looks like this uncovered usage that wouldn't have worked when using other filesystems. I've added a test for `SAMFileWriterFactory` that uses jimfs, and updated the code to be able to handle `Path` objects in a few more places to make this work.

magicDGS commented on src/main/java/htsjdk/samtools/StreamInflatingIndexingOutputStream.java:
```diff
@@ -26,7 +27,7 @@ public StreamInflatingIndexingOutputStream(final OutputStream s1, final File ind
             this.s2 = new PipedOutputStream();
             final PipedInputStream pin = new PipedInputStream(this.s2, Defaults.NON_ZERO_BUFFER_SIZE);
-            this.thread = new Thread(new Indexer(indexFile, pin), "BamIndexingThread");
+            this.thread = new Thread(new Indexer(indexFile.toPath(), pin), "BamIndexingThread");
```
Could it be possible to add a constructor for `Path` also here, and delegate the `File`one to that one?

jean-philippe-martin replied:
Sure!

magicDGS commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -179,6 +182,13 @@ public void testAddExtension() throws IOException {
         Assert.assertEquals(IOUtil.addExtension(p, ".ext"), IOUtil.getPath("/folder/file.ext"));
         p = IOUtil.getPath("folder/file");
         Assert.assertEquals(IOUtil.addExtension(p, ".ext"), IOUtil.getPath("folder/file.ext"));
+        FileSystem jimfs = Jimfs.newFileSystem(Configuration.unix());
```
The `FileSystem` should be closed. I guess that try-with-resources is the better option for that...

magicDGS commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -170,6 +175,22 @@ public void testFileTypeUnix(final String path, boolean expectedIsRegularFile) {
+        Assert.assertEquals(IOUtil.addExtension(p, ".ext"), IOUtil.getPath("/folder/file.ext"));
+        p = IOUtil.getPath("folder/file");
+        Assert.assertEquals(IOUtil.addExtension(p, ".ext"), IOUtil.getPath("folder/file.ext"));
+        FileSystem jimfs = Jimfs.newFileSystem(Configuration.unix());
```
The jmfs should be closed. I guess that try-with-resources is the better option for that...

jean-philippe-martin replied:
Done.

magicDGS commented:
The `Path` support in this PR is great, thanks for doing this @jean-philippe-martin! For me this is perfect, but let's see the suggestions from the Broad team.

I added two more comments that I found while reviewing the latest changes, one related with closing  jimfs and the other regarding another ctor for `Path`. Thank you for taking into account my comments!

jean-philippe-martin commented:
You're welcome @magicDGS, thank you for your helpful comments!

jean-philippe-martin commented:
If I understand correctly we're waiting for a thumbs-up from someone at Broad. @droazen or @lbergelson perhaps?

droazen commented:
I'll have a look later this week.

jean-philippe-martin commented:
@droazen gentle ping. This is needed for NIO writes.

jean-philippe-martin commented:
@droazen weekly ping. We need this to be able to make progress.

jean-philippe-martin commented:
@droazen ? @lbergelson ? @tfenne ? Is there anyone here? Hello? I can't add good cloud support support to GATK if every commit takes 25 days.

droazen commented:
@jean-philippe-martin Sorry for the delay on this review -- we've been dealing with some unexpected and urgent internal deadlines lately, as I discussed at our last group meeting, and have not had much time for htsjdk reviews. Have to ask you to be patient a bit longer, but we'll get to this.

yfarjoun commented:
this needs a review....

jean-philippe-martin commented:
@droazen - customary reminder that this code is waiting for review. I see conflicts are starting to creep in.

jean-philippe-martin commented:
@lbergelson if you start this week we may get a few rounds of review before I leave for my vacation.

jean-philippe-martin commented:
@droazen @lbergelson Customary reminder that this code is still waiting for review.

lbergelson commented on src/main/java/htsjdk/samtools/util/BinaryCodec.java:
```diff
@@ -91,29 +93,41 @@
 
     /**
-     * Constructs BinaryCodec from a file and set it's mode to writing or not
+     * Constructs BinaryCodec from a file and set its mode to writing or not
```
good catch

lbergelson commented on src/main/java/htsjdk/samtools/util/BinaryCodec.java:
```diff
@@ -91,29 +93,41 @@
-                this.outputStream = IOUtil.maybeBufferOutputStream(new FileOutputStream(file));
-                this.outputFileName = file.getName();
+                this.outputStream = IOUtil.maybeBufferOutputStream(Files.newOutputStream(path));
+                this.outputFileName = path.toString();
```
If we want to match the old behavior we should probably use `path.getFileName()` instead of `toString()`

jean-philippe-martin replied:
Done.

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -246,6 +246,15 @@ public static boolean isRegularPath(final File file) {
+     * file, or a path that could be a regular output file.
+     */
+    public static boolean isRegularPath(final Path path) {
+        return !Files.exists(path) || Files.isRegularFile(path);
```
This method is confusing to me, it seems like it should be named something else, but it mirrors what the existing method already does, so I guess we should leave it...

jean-philippe-martin replied:
Agreed. We can rename it in a later PR if there's enough support for it.

lbergelson commented on src/test/java/htsjdk/samtools/CRAMComplianceTest.java:
```diff
@@ -261,6 +269,40 @@ public void testBAMThroughCRAMRoundTrip() throws IOException, NoSuchAlgorithmExc
+        // These files are reduced versions of the CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam and human_g1k_v37.20.21.fasta
+        // files used in GATK4 tests. The first 8000 records from chr20 were extracted; from those around 80 placed but
+        // unmapped reads that contained cigar elements were removed, along with one read who's mate was on chr21.
+        // Finally all read positions were remapped to the subsetted reference file, which contains only the ~9000 bases
```
thank you for the explanation

jean-philippe-martin replied:
You're welcome!

lbergelson commented on src/test/java/htsjdk/samtools/CRAMComplianceTest.java:
```diff
@@ -299,4 +345,21 @@ private void writeRecordsToFile (
+        final SAMFileHeader samHeader) {
+
+        // NOTE: even when the input is coord-sorted, using assumePresorted=false will cause some
+        // tests to fail since it can change the order of some unmapped reads - AFAICT this is allowed
```
You're right about the ordering issue I believe.

lbergelson commented on src/test/java/htsjdk/samtools/SAMFileWriterFactoryTest.java:
```diff
@@ -150,6 +168,18 @@ private void createSmallBam(final File outputFile) {
+        // index only created if coordinate sorted
+        header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+        header.addSequence(new SAMSequenceRecord("chr1", 123));
+        final SAMFileWriter writer = factory.makeBAMWriter(header, false, outputPath);
```
this seems like it should be a try() with resources

lbergelson commented on src/test/java/htsjdk/samtools/SAMFileWriterFactoryTest.java:
```diff
@@ -150,6 +168,18 @@ private void createSmallBam(final File outputFile) {
         writer.close();
```
this should probably be a try with resources

jean-philippe-martin replied:
Changed here and below.

lbergelson commented on src/test/java/htsjdk/samtools/SAMFileWriterFactoryTest.java:
```diff
@@ -220,6 +250,28 @@ private void verifyWriterOutput(File outputFile, ReferenceSource refSource, int
+        if (verifySupplementalFiles) {
+            final Path index = SamFiles.findIndex(output);
+            final Path md5File = IOUtil.addExtension(output, ".md5");
+            Assert.assertTrue(Files.size(index) > 0);
```
Maybe extract this out as a helper method since it's duplicated.  Although there's a ton of duplicated code in these test files anyway...

jean-philippe-martin replied:
The text is there twice, but one of them uses a File and other other uses a Path. Not much there to deduplicate.

lbergelson commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -170,6 +175,23 @@ public void testFileTypeUnix(final String path, boolean expectedIsRegularFile) {
         Assert.assertEquals(IOUtil.isRegularPath(file), expectedIsRegularFile);
     }
 
+    @Test
```
Need a unit test for isRegularPath(path)

jean-philippe-martin replied:
added

jean-philippe-martin commented on src/main/java/htsjdk/samtools/BAMFileWriter.java:
```diff
@@ -75,7 +75,13 @@ protected BAMFileWriter(final OutputStream os, final File file, final int compre
     }
 
-    private void prepareToWriteAlignments() {
+    protected BAMFileWriter(final OutputStream os, final String aboluteFilename, final int compressionLevel, final DeflaterFactory deflaterFactory) {
```
thanks! Fixed.

jean-philippe-martin commented on src/main/java/htsjdk/samtools/BAMFileWriter.java:
```diff
@@ -99,17 +109,17 @@ void enableBamIndexConstruction () {
-            if (indexFile.exists()) {
-                if (!indexFile.canWrite()) {
-                    throw new SAMException("Not creating BAM index since unable to write index file " + indexFile);
+            final String indexFileBase = pathURI.toString().endsWith(BamFileIoUtils.BAM_FILE_EXTENSION) ?
```
removed toString()

jean-philippe-martin commented on src/main/java/htsjdk/samtools/BinaryBAMIndexWriter.java:
```diff
@@ -59,6 +60,24 @@ public BinaryBAMIndexWriter(final int nRef, final File output) {
     }
 
     /**
+     * constructor
```
Removing the word "constructor" (I think that's what you mean?)

jean-philippe-martin commented on src/main/java/htsjdk/samtools/BinaryBAMIndexWriter.java:
```diff
@@ -59,6 +60,24 @@ public BinaryBAMIndexWriter(final int nRef, final File output) {
+     * @param nRef    Number of reference sequences
+     * @param output  BAM Index output file
+     */
+    public BinaryBAMIndexWriter(final int nRef, final Path output) {
```
OK, done!

jean-philippe-martin commented on src/main/java/htsjdk/samtools/BAMIndexer.java:
```diff
@@ -56,11 +57,19 @@
+     * @param fileHeader header for the corresponding bam file
+     */
+    public BAMIndexer(final File output, final SAMFileHeader fileHeader) {
+        this(output.toPath(), fileHeader);
```
Good point. Changing it here.

jean-philippe-martin commented:
Addressed changes. For some reason I can't compile this in intellij anymore; ./gradlew test running as we speak.

jean-philippe-martin commented:
```
$ ./gradlew test
Results: SUCCESS (24183 tests, 24183 successes, 0 failures, 0 skipped)
BUILD SUCCESSFUL
```

jean-philippe-martin commented:
@lbergelson it's been a week. I'm going to start rebasing to the new master and resolving conflicts.

jean-philippe-martin commented:
Done. Pushing the new code.
```
Tests: succeeded 24455, failed 0, canceled 0, ignored 63, pending 0
All tests passed.
```

lbergelson commented:
@jean-philippe-martin It looks like there was an issue with the rebase:


```
htsjdk/src/main/java/htsjdk/samtools/SAMFileWriterFactory.java:348: error: ')' expected
                          IOUtil.addExtension(outputPath, ".md5"), samFlagFieldOutput)
```

Good to merge once that's resolved and test pass!

jean-philippe-martin commented:
Oh indeed, I forgot to add that fix to the commit. Good on Travis! Fixed now. I don't have a "merge" button here (on GitHub) so @lbergelson I'll let you do the merging once Travis comes up green (unless you want me to do that on the command line? I seem to remember some prohibition against that).

lbergelson commented:
@jean-philippe-martin Your long nightmare is finally over!

jean-philippe-martin commented:
Champagne! Thank you Louis! We can celebrate the submission of my longest-ever pull request. From April 3 to September 1st, so two days short of 5 months long! 
# 🎉
